openvswitch-2.13.3/vswitchd/ovs-vswitchd.conf.db.5
'\" p
.\" -*- nroff -*-
.TH "ovs-vswitchd.conf.db" 5 " DB Schema 8.2.0" "Open vSwitch 2.13.3" "Open vSwitch Manual"
.fp 5 L CR              \\" Make fixed-width font available as \\fL.
.de TQ
.  br
.  ns
.  TP "\\$1"
..
.de ST
.  PP
.  RS -0.15in
.  I "\\$1"
.  RE
..
.SH NAME
ovs-vswitchd\[char46]conf\[char46]db \- Open_vSwitch database schema
.PP
.PP
.PP
.PP
A database with this schema holds the configuration for one Open vSwitch daemon\[char46] The top-level configuration for the daemon is the \fBOpen_vSwitch\fR table, which must have exactly one record\[char46] Records in other tables are significant only when they can be reached directly or indirectly from the \fBOpen_vSwitch\fR table\[char46] Records that are not reachable from the \fBOpen_vSwitch\fR table are automatically deleted from the database, except for records in a few distinguished ``root set\(cq\(cq tables\[char46]
.SS "Common Columns"
.PP
.PP
Most tables contain two special columns, named \fBother_config\fR and \fBexternal_ids\fR\[char46] These columns have the same form and purpose each place that they appear, so we describe them here to save space later\[char46]
.RS
.TP
\fBother_config\fR: map of string-string pairs
Key-value pairs for configuring rarely used features\[char46] Supported keys, along with the forms taken by their values, are documented individually for each table\[char46]
.IP
A few tables do not have \fBother_config\fR columns because no key-value pairs have yet been defined for them\[char46]
.TP
\fBexternal_ids\fR: map of string-string pairs
Key-value pairs for use by external frameworks that integrate with Open vSwitch, rather than by Open vSwitch itself\[char46] System integrators should either use the Open vSwitch development mailing list to coordinate on common key-value definitions, or choose key names that are likely to be unique\[char46] In some cases, where key-value pairs have been defined that are likely to be widely useful, they are documented individually for each table\[char46]
.RE
.SH "TABLE SUMMARY"
.PP
The following list summarizes the purpose of each of the tables in the
\fBOpen_vSwitch\fR database.  Each table is described in more detail on a later
page.
.IP "Table" 1in
Purpose
.TQ 1in
\fBOpen_vSwitch\fR
Open vSwitch configuration\[char46]
.TQ 1in
\fBBridge\fR
Bridge configuration\[char46]
.TQ 1in
\fBPort\fR
Port configuration\[char46]
.TQ 1in
\fBInterface\fR
One physical network device in a Port\[char46]
.TQ 1in
\fBFlow_Table\fR
OpenFlow table configuration
.TQ 1in
\fBQoS\fR
Quality of Service configuration
.TQ 1in
\fBQueue\fR
QoS output queue\[char46]
.TQ 1in
\fBMirror\fR
Port mirroring\[char46]
.TQ 1in
\fBController\fR
OpenFlow controller configuration\[char46]
.TQ 1in
\fBManager\fR
OVSDB management connection\[char46]
.TQ 1in
\fBNetFlow\fR
NetFlow configuration\[char46]
.TQ 1in
\fBDatapath\fR
Datapath configuration\[char46]
.TQ 1in
\fBCT_Zone\fR
CT_Zone configuration\[char46]
.TQ 1in
\fBCT_Timeout_Policy\fR
CT_Timeout_Policy configuration\[char46]
.TQ 1in
\fBSSL\fR
SSL configuration\[char46]
.TQ 1in
\fBsFlow\fR
sFlow configuration\[char46]
.TQ 1in
\fBIPFIX\fR
IPFIX configuration\[char46]
.TQ 1in
\fBFlow_Sample_Collector_Set\fR
Flow_Sample_Collector_Set configuration\[char46]
.TQ 1in
\fBAutoAttach\fR
AutoAttach configuration\[char46]
.bp
.SH "Open_vSwitch TABLE"
.PP
Configuration for an Open vSwitch daemon\[char46] There must be exactly one record in the \fBOpen_vSwitch\fR table\[char46]
.SS "Summary:
.TQ .25in
\fIConfiguration:\fR
.RS .25in
.TQ 2.75in
\fBdatapaths\fR
map of string-\fBDatapath\fR pairs
.TQ 2.75in
\fBbridges\fR
set of \fBBridge\fRs
.TQ 2.75in
\fBssl\fR
optional \fBSSL\fR
.TQ 2.75in
\fBexternal_ids : system-id\fR
optional string
.TQ 2.75in
\fBexternal_ids : xs-system-uuid\fR
optional string
.TQ 2.75in
\fBexternal_ids : hostname\fR
optional string
.TQ 2.75in
\fBexternal_ids : rundir\fR
optional string
.TQ 2.75in
\fBother_config : stats-update-interval\fR
optional string, containing an integer, at least 5,000
.TQ 2.75in
\fBother_config : flow-restore-wait\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : flow-limit\fR
optional string, containing an integer, at least 0
.TQ 2.75in
\fBother_config : max-idle\fR
optional string, containing an integer, at least 500
.TQ 2.75in
\fBother_config : max-revalidator\fR
optional string, containing an integer, at least 100
.TQ 2.75in
\fBother_config : min-revalidate-pps\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : hw-offload\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : tc-policy\fR
optional string, one of \fBnone\fR, \fBskip_hw\fR, or \fBskip_sw\fR
.TQ 2.75in
\fBother_config : dpdk-init\fR
optional string, one of \fBfalse\fR, \fBtrue\fR, or \fBtry\fR
.TQ 2.75in
\fBother_config : dpdk-lcore-mask\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : pmd-cpu-mask\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-alloc-mem\fR
optional string, containing an integer, at least 0
.TQ 2.75in
\fBother_config : dpdk-socket-mem\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-socket-limit\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-hugepage-dir\fR
optional string
.TQ 2.75in
\fBother_config : dpdk-extra\fR
optional string
.TQ 2.75in
\fBother_config : vhost-sock-dir\fR
optional string
.TQ 2.75in
\fBother_config : vhost-iommu-support\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : vhost-postcopy-support\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : per-port-memory\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : tx-flush-interval\fR
optional string, containing an integer, in range 0 to 1,000,000
.TQ 2.75in
\fBother_config : pmd-perf-metrics\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : smc-enable\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : pmd-rxq-assign\fR
optional string, either \fBcycles\fR or \fBroundrobin\fR
.TQ 2.75in
\fBother_config : n-handler-threads\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : n-revalidator-threads\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : emc-insert-inv-prob\fR
optional string, containing an integer, in range 0 to 4,294,967,295
.TQ 2.75in
\fBother_config : vlan-limit\fR
optional string, containing an integer, at least 0
.TQ 2.75in
\fBother_config : bundle-idle-timeout\fR
optional string, containing an integer, at least 1
.TQ 2.75in
\fBother_config : offload-rebalance\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : pmd-auto-lb\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.75in
\fBother_config : pmd-auto-lb-rebal-interval\fR
optional string, containing an integer, in range 0 to 20,000
.TQ 2.75in
\fBother_config : userspace-tso-enable\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.RE
.TQ .25in
\fIStatus:\fR
.RS .25in
.TQ 2.75in
\fBnext_cfg\fR
integer
.TQ 2.75in
\fBcur_cfg\fR
integer
.TQ 2.75in
\fBdpdk_initialized\fR
boolean
.TQ .25in
\fIStatistics:\fR
.RS .25in
.TQ 2.50in
\fBother_config : enable-statistics\fR
optional string, either \fBtrue\fR or \fBfalse\fR
.TQ 2.50in
\fBstatistics : cpu\fR
optional string, containing an integer, at least 1
.TQ 2.50in
\fBstatistics : load_average\fR
optional string
.TQ 2.50in
\fBstatistics : memory\fR
optional string
.TQ 2.50in
\fBstatistics : process_NAME\fR
optional string
.TQ 2.50in
\fBstatistics : file_systems\fR
optional string
.RE
.RE
.TQ .25in
\fIVersion Reporting:\fR
.RS .25in
.TQ 2.75in
\fBovs_version\fR
optional string
.TQ 2.75in
\fBdb_version\fR
optional string
.TQ 2.75in
\fBsystem_type\fR
optional string
.TQ 2.75in
\fBsystem_version\fR
optional string
.TQ 2.75in
\fBdpdk_version\fR
optional string
.RE
.TQ .25in
\fICapabilities:\fR
.RS .25in
.TQ 2.75in
\fBdatapath_types\fR
set of strings
.TQ 2.75in
\fBiface_types\fR
set of strings
.RE
.TQ .25in
\fIDatabase Configuration:\fR
.RS .25in
.TQ 2.75in
\fBmanager_options\fR
set of \fBManager\fRs
.RE
.TQ .25in
\fIIPsec:\fR
.RS .25in
.TQ 2.75in
\fBother_config : private_key\fR
optional string
.TQ 2.75in
\fBother_config : certificate\fR
optional string
.TQ 2.75in
\fBother_config : ca_cert\fR
optional string
.TQ .25in
\fIPlaintext Tunnel Policy:\fR
.RS .25in
.TQ 2.50in
\fBother_config : ipsec_skb_mark\fR
optional string
.RE
.RE
.TQ .25in
\fICommon Columns:\fR
.RS .25in
.TQ 2.75in
\fBother_config\fR
map of string-string pairs
.TQ 2.75in
\fBexternal_ids\fR
map of string-string pairs
.RE
.SS "Details:
.ST "Configuration:"
.PP
.IP "\fBdatapaths\fR: map of string-\fBDatapath\fR pairs"
Map of datapath types to datapaths\[char46] The \fBdatapath_type\fR column of the \fBBridge\fR table is used as a key for this map\[char46] The value points to a row in the \fBDatapath\fR table\[char46]
.IP "\fBbridges\fR: set of \fBBridge\fRs"
Set of bridges managed by the daemon\[char46]
.IP "\fBssl\fR: optional \fBSSL\fR"
SSL used globally by the daemon\[char46]
.IP "\fBexternal_ids : system-id\fR: optional string"
A unique identifier for the Open vSwitch\(cqs physical host\[char46] The form of the identifier depends on the type of the host\[char46] On a Citrix XenServer, this will likely be the same as \fBexternal_ids:xs-system-uuid\fR\[char46]
.IP "\fBexternal_ids : xs-system-uuid\fR: optional string"
The Citrix XenServer universally unique identifier for the physical host as displayed by \fBxe host\-list\fR\[char46]
.IP "\fBexternal_ids : hostname\fR: optional string"
The hostname for the host running Open vSwitch\[char46] This is a fully qualified domain name since version 2\[char46]6\[char46]2\[char46]
.IP "\fBexternal_ids : rundir\fR: optional string"
In Open vSwitch 2\[char46]8 and later, the run directory of the running Open vSwitch daemon\[char46] This directory is used for runtime state such as control and management sockets\[char46] The value of \fBother_config:vhost-sock-dir\fR is relative to this directory\[char46]
.IP "\fBother_config : stats-update-interval\fR: optional string, containing an integer, at least 5,000"
Interval for updating statistics to the database, in milliseconds\[char46] This option will affect the update of the \fBstatistics\fR column in the following tables: \fBPort\fR, \fBInterface
\fR, \fBMirror\fR\[char46]
.IP
Default value is 5000 ms\[char46]
.IP
Getting statistics more frequently can be achieved via OpenFlow\[char46]
.IP "\fBother_config : flow-restore-wait\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
When \fBovs\-vswitchd\fR starts up, it has an empty flow table and therefore it handles all arriving packets in its default fashion according to its configuration, by dropping them or sending them to an OpenFlow controller or switching them as a standalone switch\[char46] This behavior is ordinarily desirable\[char46] However, if \fBovs\-vswitchd\fR is restarting as part of a ``hot-upgrade,\(cq\(cq then this leads to a relatively long period during which packets are mishandled\[char46]
.IP
This option allows for improvement\[char46] When \fBovs\-vswitchd\fR starts with this value set as \fBtrue\fR, it will neither flush or expire previously set datapath flows nor will it send and receive any packets to or from the datapath\[char46] When this value is later set to \fBfalse\fR, \fBovs\-vswitchd\fR will start receiving packets from the datapath and re-setup the flows\[char46]
.IP
Additionally, \fBovs\-vswitchd\fR is prevented from connecting to controllers when this value is set to \fBtrue\fR\[char46] This prevents controllers from making changes to the flow table in the middle of flow restoration, which could result in undesirable intermediate states\[char46] Once this value has been set to \fBfalse\fR and the desired flow state has been restored, \fBovs\-vswitchd\fR will be able to reconnect to controllers and process any new flow table modifications\[char46]
.IP
Thus, with this option, the procedure for a hot-upgrade of \fBovs\-vswitchd\fR becomes roughly the following:
.RS
.IP 1. .4in
Stop \fBovs\-vswitchd\fR\[char46]
.IP 2. .4in
Set \fBother_config:flow-restore-wait\fR to \fBtrue\fR\[char46]
.IP 3. .4in
Start \fBovs\-vswitchd\fR\[char46]
.IP 4. .4in
Use \fBovs\-ofctl\fR (or some other program, such as an OpenFlow controller) to restore the OpenFlow flow table to the desired state\[char46]
.IP 5. .4in
Set \fBother_config:flow-restore-wait\fR to \fBfalse\fR (or remove it entirely from the database)\[char46]
.RE
.IP
The \fBovs\-ctl\fR\(cqs ``restart\(cq\(cq and ``force-reload-kmod\(cq\(cq functions use the above config option during hot upgrades\[char46]
.IP "\fBother_config : flow-limit\fR: optional string, containing an integer, at least 0"
The maximum number of flows allowed in the datapath flow table\[char46] Internally OVS will choose a flow limit which will likely be lower than this number, based on real time network conditions\[char46] Tweaking this value is discouraged unless you know exactly what you\(cqre doing\[char46]
.IP
The default is 200000\[char46]
.IP "\fBother_config : max-idle\fR: optional string, containing an integer, at least 500"
The maximum time (in ms) that idle flows will remain cached in the datapath\[char46] Internally OVS will check the validity and activity for datapath flows regularly and may expire flows quicker than this number, based on real time network conditions\[char46] Tweaking this value is discouraged unless you know exactly what you\(cqre doing\[char46]
.IP
The default is 10000\[char46]
.IP "\fBother_config : max-revalidator\fR: optional string, containing an integer, at least 100"
The maximum time (in ms) that revalidator threads will wait before executing flow revalidation\[char46] Note that this is maximum allowed value\[char46] Actual timeout used by OVS is minimum of max-idle and max-revalidator values\[char46] Tweaking this value is discouraged unless you know exactly what you\(cqre doing\[char46]
.IP
The default is 500\[char46]
.IP "\fBother_config : min-revalidate-pps\fR: optional string, containing an integer, at least 1"
Set minimum pps that flow must have in order to be revalidated when revalidation duration exceeds half of max-revalidator config variable\[char46]
.IP
The default is 5\[char46]
.IP "\fBother_config : hw-offload\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
Set this value to \fBtrue\fR to enable netdev flow offload\[char46]
.IP
The default value is \fBfalse\fR\[char46] Changing this value requires restarting the daemon
.IP
Currently Open vSwitch supports hardware offloading on Linux systems\[char46] On other systems, this value is ignored\[char46] This functionality is considered \(cqexperimental\(cq\[char46] Depending on which OpenFlow matches and actions are configured, which kernel version is used, and what hardware is available, Open vSwitch may not be able to offload functionality to hardware\[char46]
.IP
In order to dump HW offloaded flows use \fBovs\-appctl dpctl/dump\-flows\fR, \fBovs\-dpctl\fR doesn\(cqt support this functionality\[char46] See ovs-vswitchd(8) for details\[char46]
.IP "\fBother_config : tc-policy\fR: optional string, one of \fBnone\fR, \fBskip_hw\fR, or \fBskip_sw\fR"
Specified the policy used with HW offloading\[char46] Options:
.RS
.TP
\fBnone\fR
Add software rule and offload rule to HW\[char46]
.TP
\fBskip_sw\fR
Offload rule to HW only\[char46]
.TP
\fBskip_hw\fR
Add software rule without offloading rule to HW\[char46]
.RE
.IP
This is only relevant if \fBother_config:hw-offload\fR is enabled\[char46]
.IP
The default value is \fBnone\fR\[char46]
.IP "\fBother_config : dpdk-init\fR: optional string, one of \fBfalse\fR, \fBtrue\fR, or \fBtry\fR"
Set this value to \fBtrue\fR or \fBtry\fR to enable runtime support for DPDK ports\[char46] The vswitch must have compile-time support for DPDK as well\[char46]
.IP
A value of \fBtrue\fR will cause the ovs-vswitchd process to abort if DPDK cannot be initialized\[char46] A value of \fBtry\fR will allow the ovs-vswitchd process to continue running even if DPDK cannot be initialized\[char46]
.IP
The default value is \fBfalse\fR\[char46] Changing this value requires restarting the daemon
.IP
If this value is \fBfalse\fR at startup, any dpdk ports which are configured in the bridge will fail due to memory errors\[char46]
.IP "\fBother_config : dpdk-lcore-mask\fR: optional string, containing an integer, at least 1"
Specifies the CPU cores where dpdk lcore threads should be spawned\[char46] The DPDK lcore threads are used for DPDK library tasks, such as library internal message processing, logging, etc\[char46] Value should be in the form of a hex string (so \(cq0x123\(cq) similar to the \(cqtaskset\(cq mask input\[char46]
.IP
The lowest order bit corresponds to the first CPU core\[char46] A set bit means the corresponding core is available and an lcore thread will be created and pinned to it\[char46] If the input does not cover all cores, those uncovered cores are considered not set\[char46]
.IP
For performance reasons, it is best to set this to a single core on the system, rather than allow lcore threads to float\[char46]
.IP
If not specified, the value will be determined by choosing the lowest CPU core from initial cpu affinity list\[char46] Otherwise, the value will be passed directly to the DPDK library\[char46]
.IP "\fBother_config : pmd-cpu-mask\fR: optional string"
Specifies CPU mask for setting the cpu affinity of PMD (Poll Mode Driver) threads\[char46] Value should be in the form of hex string, similar to the dpdk EAL \(cq-c COREMASK\(cq option input or the \(cqtaskset\(cq mask input\[char46]
.IP
The lowest order bit corresponds to the first CPU core\[char46] A set bit means the corresponding core is available and a pmd thread will be created and pinned to it\[char46] If the input does not cover all cores, those uncovered cores are considered not set\[char46]
.IP
If not specified, one pmd thread will be created for each numa node and pinned to any available core on the numa node by default\[char46]
.IP "\fBother_config : dpdk-alloc-mem\fR: optional string, containing an integer, at least 0"
Specifies the amount of memory to preallocate from the hugepage pool, regardless of socket\[char46] It is recommended that dpdk-socket-mem is used instead\[char46]
.IP "\fBother_config : dpdk-socket-mem\fR: optional string"
Specifies the amount of memory to preallocate from the hugepage pool, on a per-socket basis\[char46]
.IP
The specifier is a comma-separated string, in ascending order of CPU socket\[char46] E\[char46]g\[char46] On a four socket system 1024,0,2048 would set socket 0 to preallocate 1024MB, socket 1 to preallocate 0MB, socket 2 to preallocate 2048MB and socket 3 (no value given) to preallocate 0MB\[char46]
.IP
If dpdk-socket-mem and dpdk-alloc-mem are not specified, dpdk-socket-mem will be used and the default value is 1024 for each numa node\[char46] If dpdk-socket-mem and dpdk-alloc-mem are specified at same time, dpdk-socket-mem will be used as default\[char46] Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : dpdk-socket-limit\fR: optional string"
Limits the maximum amount of memory that can be used from the hugepage pool, on a per-socket basis\[char46]
.IP
The specifier is a comma-separated list of memory limits per socket\[char46] \fB0\fR will disable the limit for a particular socket\[char46]
.IP
If not specified, OVS will configure limits equal to the amount of preallocated memory specified by \fBother_config:dpdk-socket-mem\fR or \fB\-\-socket\-mem\fR in \fBother_config:dpdk-extra\fR\[char46] If none of the above options specified or \fB\-\-legacy\-mem\fR provided in \fBother_config:dpdk-extra\fR, limits will not be applied\[char46] Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : dpdk-hugepage-dir\fR: optional string"
Specifies the path to the hugetlbfs mount point\[char46]
.IP
If not specified, this will be guessed by the DPDK library (default is /dev/hugepages)\[char46] Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : dpdk-extra\fR: optional string"
Specifies additional eal command line arguments for DPDK\[char46]
.IP
The default is empty\[char46] Changing this value requires restarting the daemon
.IP "\fBother_config : vhost-sock-dir\fR: optional string"
Specifies a relative path from \fBexternal_ids:rundir\fR to the vhost-user unix domain socket files\[char46] If this value is unset, the sockets are put directly in \fBexternal_ids:rundir\fR\[char46]
.IP
Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : vhost-iommu-support\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
vHost IOMMU is a security feature, which restricts the vhost memory that a virtio device may access\[char46] vHost IOMMU support is disabled by default, due to a bug in QEMU implementations of the vhost REPLY_ACK protocol, (on which vHost IOMMU relies) prior to v2\[char46]9\[char46]1\[char46] Setting this value to \fBtrue\fR enables vHost IOMMU support for vHost User Client ports in OvS-DPDK, starting from DPDK v17\[char46]11\[char46]
.IP
Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : vhost-postcopy-support\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
vHost post-copy is a feature which allows switching live migration of VM attached to dpdkvhostuserclient port to post-copy mode if default pre-copy migration can not be converged or takes too long to converge\[char46] Setting this value to \fBtrue\fR enables vHost post-copy support for all dpdkvhostuserclient ports\[char46] Available starting from DPDK v18\[char46]11 and QEMU 2\[char46]12\[char46]
.IP
Changing this value requires restarting the daemon\[char46]
.IP "\fBother_config : per-port-memory\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
By default OVS DPDK uses a shared memory model wherein devices that have the same MTU and socket values can share the same mempool\[char46] Setting this value to \fBtrue\fR changes this behaviour\[char46] Per port memory allow DPDK devices to use private memory per device\[char46] This can provide greater transparency as regards memory usage but potentially at the cost of greater memory requirements\[char46]
.IP
Changing this value requires restarting the daemon if dpdk-init has already been set to true\[char46]
.IP "\fBother_config : tx-flush-interval\fR: optional string, containing an integer, in range 0 to 1,000,000"
Specifies the time in microseconds that a packet can wait in output batch for sending i\[char46]e\[char46] amount of time that packet can spend in an intermediate output queue before sending to netdev\[char46] This option can be used to configure balance between throughput and latency\[char46] Lower values decreases latency while higher values may be useful to achieve higher performance\[char46]
.IP
Defaults to 0 i\[char46]e\[char46] instant packet sending (latency optimized)\[char46]
.IP "\fBother_config : pmd-perf-metrics\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
Enables recording of detailed PMD performance metrics for analysis and trouble-shooting\[char46] This can have a performance impact in the order of 1%\[char46]
.IP
Defaults to false but can be changed at any time\[char46]
.IP "\fBother_config : smc-enable\fR: optional string, either \fBtrue\fR or \fBfalse\fR"
Signature match cache or SMC is a cache between EMC and megaflow cache\[char46] It does not store the full key of the flow, so it is more memory efficient comparing to EMC cache\[char46] SMC is especially useful when flow count is larger than EMC capacity\[char46]
.IP
Defaults to false but can be changed at any time\[char46]

openvswitch-2.13.3/vswitchd/ovs-vswitchd.c
/* Copyright (c) 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#ifdef HAVE_MLOCKALL
#include <sys/mman.h>
#endif

#include "bridge.h"
#include "command-line.h"
#include "compiler.h"
#include "daemon.h"
#include "dirs.h"
#include "dpif.h"
#include "dummy.h"
#include "fatal-signal.h"
#include "memory.h"
#include "netdev.h"
#include "openflow/openflow.h"
#include "ovsdb-idl.h"
#include "ovs-rcu.h"
#include "ovs-router.h"
#include "ovs-thread.h"
#include "openvswitch/poll-loop.h"
#include "simap.h"
#include "stream-ssl.h"
#include "stream.h"
#include "svec.h"
#include "timeval.h"
#include "unixctl.h"
#include "util.h"
#include "openvswitch/vconn.h"
#include "openvswitch/vlog.h"
#include "lib/vswitch-idl.h"
#include "lib/dns-resolve.h"

VLOG_DEFINE_THIS_MODULE(vswitchd);

/* --mlockall: If set, locks all process memory into physical RAM, preventing
 * the kernel from paging any of its memory to disk. */
static bool want_mlockall;

static unixctl_cb_func ovs_vswitchd_exit;

static char *parse_options(int argc, char *argv[], char **unixctl_path);
OVS_NO_RETURN static void usage(void);

struct ovs_vswitchd_exit_args {
    bool *exiting;
    bool *cleanup;
};

int
main(int argc, char *argv[])
{   
    char *unixctl_path = NULL;
    struct unixctl_server *unixctl;
    char *remote;
    bool exiting, cleanup;
    struct ovs_vswitchd_exit_args exit_args = {&exiting, &cleanup};
    int retval;

    set_program_name(argv[0]);//设置程序名称，版本，编译日期等
    ovsthread_id_init();

    dns_resolve_init(true);
    ovs_cmdl_proctitle_init(argc, argv);//复制出输入的参数列表到新的存储中，让argv指向这块内存
    service_start(&argc, &argv);//注册回调和服务管理器出现故障错误时候操作的配置
    remote = parse_options(argc, argv, &unixctl_path);//解析参数，其中unixctl_path存储着sock,作为接收外部控制命令的渠道，而remote存储连接到ovsdb信息，也就是连接到配置数据库的sock名
    fatal_ignore_sigpipe();

    daemonize_start(true);//开始守候进程模式

    if (want_mlockall) {
#ifdef HAVE_MLOCKALL
        if (mlockall(MCL_CURRENT | MCL_FUTURE)) {
            VLOG_ERR("mlockall failed: %s", ovs_strerror(errno));
        } else {
            set_memory_locked();
        }
#else
        VLOG_ERR("mlockall not supported on this system");
#endif
    }

    retval = unixctl_server_create(unixctl_path, &unixctl);//创建一个unixctl_server(存放在unixctl)，并监听在unixctl_path指定的punix路径，该路径作为ovs-appctl发送命令给ovsd的通道
    if (retval) {
        exit(EXIT_FAILURE);
    }
    unixctl_command_register("exit", "[--cleanup]", 0, 1,
                             ovs_vswitchd_exit, &exit_args);

    bridge_init(remote);//初始化这个bridge
    free(remote);

    exiting = false;
    cleanup = false;
    //main函数里面的一直循环的循环体
    while (!exiting) {
        memory_run();//运行内存监视器，客户端调用memory_should_report()。此函数以及该模块的接口的剩余部分，仅被一个线程调用。
        if (memory_should_report()) {
            struct simap usage;

            simap_init(&usage);
            bridge_get_memory_usage(&usage);
            memory_report(&usage);
            simap_destroy(&usage);
        }
        bridge_run();//其中bridge_run就是初始化数据库中已经创建的虚拟网桥 ////处理与controller的交互，ovs-ofctl命令
        unixctl_server_run(unixctl);//从unixctl指定的server获取ovs-appctl发送的命令并执行 ////处理ovs-vsctl命令
        netdev_run();//在内核中添加虚拟网卡，也就是初始化数据库中的虚拟网卡

        memory_wait();
        bridge_wait();
        unixctl_server_wait(unixctl);
        netdev_wait();
        if (exiting) {
            poll_immediate_wake();
        }
        poll_block();//阻塞知道被poll_fd_wait()注册的事件发生或者超时
        if (should_service_stop()) {
            exiting = true;
        }
    }
    bridge_exit(cleanup);
    unixctl_server_destroy(unixctl);
    service_stop();
    vlog_disable_async();
    ovsrcu_exit();
    dns_resolve_destroy();

    return 0;
}

static char *
parse_options(int argc, char *argv[], char **unixctl_pathp)
{
    enum {
        OPT_PEER_CA_CERT = UCHAR_MAX + 1,
        OPT_MLOCKALL,
        OPT_UNIXCTL,
        VLOG_OPTION_ENUMS,
        OPT_BOOTSTRAP_CA_CERT,
        OPT_ENABLE_DUMMY,
        OPT_DISABLE_SYSTEM,
        OPT_DISABLE_SYSTEM_ROUTE,
        DAEMON_OPTION_ENUMS,
        OPT_DPDK,
        SSL_OPTION_ENUMS,
        OPT_DUMMY_NUMA,
    };
    static const struct option long_options[] = {
        {"help",        no_argument, NULL, 'h'},
        {"version",     no_argument, NULL, 'V'},
        {"mlockall",    no_argument, NULL, OPT_MLOCKALL},
        {"unixctl",     required_argument, NULL, OPT_UNIXCTL},
        DAEMON_LONG_OPTIONS,
        VLOG_LONG_OPTIONS,
        STREAM_SSL_LONG_OPTIONS,
        {"peer-ca-cert", required_argument, NULL, OPT_PEER_CA_CERT},
        {"bootstrap-ca-cert", required_argument, NULL, OPT_BOOTSTRAP_CA_CERT},
        {"enable-dummy", optional_argument, NULL, OPT_ENABLE_DUMMY},
        {"disable-system", no_argument, NULL, OPT_DISABLE_SYSTEM},
        {"disable-system-route", no_argument, NULL, OPT_DISABLE_SYSTEM_ROUTE},
        {"dpdk", optional_argument, NULL, OPT_DPDK},
        {"dummy-numa", required_argument, NULL, OPT_DUMMY_NUMA},
        {NULL, 0, NULL, 0},
    };
    char *short_options = ovs_cmdl_long_options_to_short_options(long_options);

    for (;;) {
        int c;

        c = getopt_long(argc, argv, short_options, long_options, NULL);
        if (c == -1) {
            break;
        }

        switch (c) {
        case 'h':
            usage();

        case 'V':
            ovs_print_version(0, 0);
            print_dpdk_version();
            exit(EXIT_SUCCESS);

        case OPT_MLOCKALL:
            want_mlockall = true;
            break;

        case OPT_UNIXCTL:
            *unixctl_pathp = optarg;
            break;

        VLOG_OPTION_HANDLERS
        DAEMON_OPTION_HANDLERS
        STREAM_SSL_OPTION_HANDLERS

        case OPT_PEER_CA_CERT:
            stream_ssl_set_peer_ca_cert_file(optarg);
            break;

        case OPT_BOOTSTRAP_CA_CERT:
            stream_ssl_set_ca_cert_file(optarg, true);
            break;

        case OPT_ENABLE_DUMMY:
            dummy_enable(optarg);
            break;

        case OPT_DISABLE_SYSTEM:
            dp_blacklist_provider("system");
            break;

        case OPT_DISABLE_SYSTEM_ROUTE:
            ovs_router_disable_system_routing_table();
            break;

        case '?':
            exit(EXIT_FAILURE);

        case OPT_DPDK:
            ovs_fatal(0, "Using --dpdk to configure DPDK is not supported.");
            break;

        case OPT_DUMMY_NUMA:
            ovs_numa_set_dummy(optarg);
            break;

        default:
            abort();
        }
    }
    free(short_options);

    argc -= optind;
    argv += optind;

    switch (argc) {
    case 0:
        return xasprintf("unix:%s/db.sock", ovs_rundir());

    case 1:
        return xstrdup(argv[0]);

    default:
        VLOG_FATAL("at most one non-option argument accepted; "
                   "use --help for usage");
    }
}

static void
usage(void)
{
    printf("%s: Open vSwitch daemon\n"
           "usage: %s [OPTIONS] [DATABASE]\n"
           "where DATABASE is a socket on which ovsdb-server is listening\n"
           "      (default: \"unix:%s/db.sock\").\n",
           program_name, program_name, ovs_rundir());
    stream_usage("DATABASE", true, false, true);
    daemon_usage();
    vlog_usage();
    printf("\nDPDK options:\n"
           "Configuration of DPDK via command-line is removed from this\n"
           "version of Open vSwitch. DPDK is configured through ovsdb.\n"
          );
    printf("\nOther options:\n"
           "  --unixctl=SOCKET          override default control socket name\n"
           "  -h, --help                display this help message\n"
           "  -V, --version             display version information\n");
    exit(EXIT_SUCCESS);
}


//优雅的退出
static void
ovs_vswitchd_exit(struct unixctl_conn *conn, int argc,
                  const char *argv[], void *exit_args_)
{
    struct ovs_vswitchd_exit_args *exit_args = exit_args_;
    *exit_args->exiting = true;
    *exit_args->cleanup = argc == 2 && !strcmp(argv[1], "--cleanup");
    unixctl_command_reply(conn, NULL);
}
